File: frontend/src/api/items.ts
import { apiFetch } from "./client";
import type { Creator } from "./itemFields";

/**
 * Item.data is opaque JSON on the wire; this typed view captures the
 * shape the form + detail panel expect. Unknown keys still round-trip
 * cleanly because `data` extends Record<string, unknown>.
 *
 * Tags are *not* in `data` — they live in the normalised `tags`
 * table and ride alongside the item as `tag_ids`. See api/tags.ts.
 */
export interface ItemData extends Record<string, unknown> {
  title?: string;
  creators?: Creator[];
  date?: string;
  abstractNote?: string;
  url?: string;
  extra?: string;
  note?: string;
}

export interface Item {
  id: string;
  space_id: string;
  item_type: string;
  data: ItemData;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  collection_ids: string[];
  tag_ids: string[];
  /**
   * Reached this listing through an inheritance link rather than living
   * in the space that was asked for. Read-only here — `space_id` says
   * where it actually lives. Only set by the per-space listing; the
   * single-item routes have no "here" to be inherited into.
   */
  is_inherited?: boolean;
}

export type ItemStatus = "active" | "trashed" | "all";
export type ItemSort = "updated" | "created" | "title" | "type";
export type SortDirection = "asc" | "desc";
export type SearchScope =
  | "title"
  | "creators"
  | "abstract"
  | "extra"
  | "fulltext";

export const ALL_SEARCH_SCOPES: SearchScope[] = [
  "title",
  "creators",
  "abstract",
  "extra",
  "fulltext",
];

export const SEARCH_SCOPE_LABELS: Record<SearchScope, string> = {
  title: "Title",
  creators: "Creators",
  abstract: "Abstract",
  extra: "Extra",
  fulltext: "PDF body",
};

/**
 * What to do with standards this space has pinned a revision of.
 *
 * "pinned" (the server default) shows the chosen revision and hides its
 * siblings; "all" reveals them. Standards with no pin are unaffected.
 */
export type RevisionFilter = "pinned" | "all";

export interface ItemCreate {
  item_type: string;
  data?: ItemData;
}

export interface ItemUpdate {
  item_type?: string;
  data?: ItemData;
}

export interface ListItemsResponse {
  items: Item[];
  /** Total matches for the current filter set, independent of paging. */
  total: number;
}

export function listItems(
  slug: string,
  opts: {
    limit?: number;
    offset?: number;
    q?: string;
    tags?: string[];
    status?: ItemStatus;
    sort?: ItemSort;
    direction?: SortDirection;
    collection?: string;
    scope?: SearchScope[];
    revisions?: RevisionFilter;
  } = {},
): Promise<ListItemsResponse> {
  const params = new URLSearchParams();
  if (opts.limit != null) params.set("limit", String(opts.limit));
  if (opts.offset != null) params.set("offset", String(opts.offset));
  if (opts.q && opts.q.trim()) params.set("q", opts.q.trim());
  if (opts.status && opts.status !== "active") params.set("status", opts.status);
  if (opts.sort && opts.sort !== "updated") params.set("sort", opts.sort);
  if (opts.direction && opts.direction !== "desc") {
    params.set("direction", opts.direction);
  }
  if (opts.tags) {
    for (const t of opts.tags) {
      const trimmed = t.trim();
      if (trimmed) params.append("tag", trimmed);
    }
  }
  // Only emit scope= when the caller has narrowed past the default
  // (all four). Equal-set short-circuit keeps URLs tidy.
  if (
    opts.scope &&
    opts.scope.length > 0 &&
    opts.scope.length < ALL_SEARCH_SCOPES.length
  ) {
    for (const s of opts.scope) params.append("scope", s);
  }
  if (opts.collection) params.set("collection", opts.collection);
  // Only when overriding: "pinned" is the server default and sending it
  // would put a redundant param in every URL and every query key.
  if (opts.revisions === "all") params.set("revisions", "all");
  const qs = params.toString();
  return apiFetch<ListItemsResponse>(
    `/api/spaces/${encodeURIComponent(slug)}/items${qs ? `?${qs}` : ""}`,
  );
}

export function getItem(id: string): Promise<Item> {
  return apiFetch<Item>(`/api/items/${encodeURIComponent(id)}`);
}

export function createItem(slug: string, payload: ItemCreate): Promise<Item> {
  return apiFetch<Item>(`/api/spaces/${encodeURIComponent(slug)}/items`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateItem(id: string, payload: ItemUpdate): Promise<Item> {
  return apiFetch<Item>(`/api/items/${encodeURIComponent(id)}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteItem(id: string): Promise<void> {
  return apiFetch<void>(`/api/items/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });
}

export function permanentDeleteItem(id: string): Promise<void> {
  return apiFetch<void>(
    `/api/items/${encodeURIComponent(id)}?permanent=true`,
    { method: "DELETE" },
  );
}

export function restoreItem(id: string): Promise<Item> {
  return apiFetch<Item>(`/api/items/${encodeURIComponent(id)}/restore`, {
    method: "POST",
  });
}

export interface CopyItemResult {
  item_id: string;
  space_id: string;
  space_slug: string;
  attachments_copied: number;
  linked_to_standard: boolean;
}

/**
 * Copy an item into another space — a real copy, bytes included.
 *
 * For a document many spaces need to *read*, subscribing to the space
 * that owns it is better: one copy, one place to fix a mistake. This is
 * for when the target genuinely wants its own.
 *
 * Notes and highlights deliberately don't come across; they belong to
 * whoever wrote them, under the visibility they chose.
 */
export function copyItem(
  id: string,
  targetSlug: string,
  opts: { includeAttachments?: boolean } = {},
): Promise<CopyItemResult> {
  return apiFetch<CopyItemResult>(`/api/items/${encodeURIComponent(id)}/copy`, {
    method: "POST",
    body: JSON.stringify({
      target_slug: targetSlug,
      include_attachments: opts.includeAttachments ?? true,
    }),
  });
}

export interface FulltextHit {
  page_number: number;
  // Server-rendered HTML — only ts_headline's <mark> tags are
  // emitted; the surrounding text is escaped by Postgres before the
  // tags are inserted. Render via dangerouslySetInnerHTML.
  snippet_html: string;
}

export interface FulltextHitsAttachment {
  attachment_id: string;
  filename: string;
  hits: FulltextHit[];
}

export function fetchFulltextHits(
  itemId: string,
  q: string,
): Promise<FulltextHitsAttachment[]> {
  const params = new URLSearchParams({ q });
  return apiFetch<FulltextHitsAttachment[]>(
    `/api/items/${encodeURIComponent(itemId)}/fulltext-hits?${params}`,
  );
}

File: backend/src/shelf/api/items.py
"""Item CRUD endpoints.

List / create / get / update / soft-delete / restore /
permanent-delete. Permissions come from the caller's role in the item's
space: viewer reads, editor writes (see auth/spaces.py).
"""

import uuid
from datetime import UTC, datetime
from enum import StrEnum
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, ConfigDict
from sqlalchemy import case, cast, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.types import String

from ..auth.deps import get_current_user
from ..auth.spaces import (
    SPACE_ROLE_EDITOR,
    SPACE_ROLE_VIEWER,
    require_space_role,
)
from ..db import get_session
from ..models import (
    Attachment,
    AttachmentPage,
    Item,
    ItemCollection,
    ItemTag,
    Space,
    Tag,
    User,
)

router = APIRouter(tags=["items"])


def _escape_ilike(s: str) -> str:
    """Quote LIKE/ILIKE wildcard chars so a user-supplied query is
    treated as a literal substring. Backslash is the escape char we
    pass via `escape="\\\\"` at call sites; chosen because it's not
    one of LIKE's metacharacters and keeps the SQL readable."""
    return s.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")


# Snippet rendering window for fulltext hits. ~60 chars on each side
# of the match keeps the snippet readable on the LibraryPage row
# without truncating short pages mid-sentence.
_SNIPPET_PAD = 60
_SNIPPET_MAX_PER_PAGE = 1


def _build_snippet(text: str, needle: str) -> str | None:
    """Return an HTML-safe snippet of `text` with the first match of
    `needle` (case-insensitive) wrapped in `<mark>…</mark>`. Other
    HTML metacharacters are escaped so we can safely render via
    dangerouslySetInnerHTML on the client."""
    from html import escape as html_escape

    if not needle:
        return None
    lower = text.lower()
    idx = lower.find(needle.lower())
    if idx == -1:
        return None
    start = max(0, idx - _SNIPPET_PAD)
    end = min(len(text), idx + len(needle) + _SNIPPET_PAD)
    pre = ("…" if start > 0 else "") + html_escape(text[start:idx])
    hit = html_escape(text[idx : idx + len(needle)])
    post = html_escape(text[idx + len(needle) : end]) + (
        "…" if end < len(text) else ""
    )
    # Collapse whitespace so multi-line PDF text wraps as a single
    # readable line in the result row.
    pre = " ".join(pre.split())
    post = " ".join(post.split())
    return f"{pre} <mark>{hit}</mark> {post}".strip()


class ItemCreate(BaseModel):
    item_type: str
    data: dict[str, Any] = {}


class ItemUpdate(BaseModel):
    item_type: str | None = None
    data: dict[str, Any] | None = None


class ItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    space_id: uuid.UUID
    item_type: str
    data: dict[str, Any]
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None
    collection_ids: list[uuid.UUID] = []
    tag_ids: list[uuid.UUID] = []


class ListItemsResponse(BaseModel):
    """Paginated items response. ``total`` is the count for the
    current filter set across the whole space — independent of the
    requested page — so the SPA can render "X items" without having
    to fetch every page just to count them."""

    items: list[ItemResponse]
    total: int


async def _attach_collection_ids(
    db: AsyncSession, items: list[Item]
) -> list[dict[str, Any]]:
    """Hydrate items with their collection_ids and tag_ids in two
    single queries (one per join). Returning bare ORM rows would force
    the response_model to figure out the joins lazily; we emit dicts
    so FastAPI can serialise without issuing an N+1.
    """
    if not items:
        return []
    ids = [it.id for it in items]
    coll_rows = await db.execute(
        select(ItemCollection.item_id, ItemCollection.collection_id).where(
            ItemCollection.item_id.in_(ids)
        )
    )
    coll_by_item: dict[uuid.UUID, list[uuid.UUID]] = {iid: [] for iid in ids}
    for item_id, coll_id in coll_rows.all():
        coll_by_item[item_id].append(coll_id)
    tag_rows = await db.execute(
        select(ItemTag.item_id, ItemTag.tag_id).where(ItemTag.item_id.in_(ids))
    )
    tag_by_item: dict[uuid.UUID, list[uuid.UUID]] = {iid: [] for iid in ids}
    for item_id, tag_id in tag_rows.all():
        tag_by_item[item_id].append(tag_id)
    return [
        {
            "id": it.id,
            "space_id": it.space_id,
            "item_type": it.item_type,
            "data": it.data,
            "created_at": it.created_at,
            "updated_at": it.updated_at,
            "deleted_at": it.deleted_at,
            "collection_ids": coll_by_item.get(it.id, []),
            "tag_ids": tag_by_item.get(it.id, []),
        }
        for it in items
    ]


class ItemStatus(StrEnum):
    active = "active"
    trashed = "trashed"
    all = "all"


class ItemSort(StrEnum):
    # Trailing underscores on the Python names dodge collisions with
    # `str.title()` and `type` (a builtin) — the wire values stay
    # "title" and "type" so the query string is unaffected.
    updated = "updated"
    created = "created"
    title_ = "title"
    type_ = "type"


class SortDirection(StrEnum):
    asc = "asc"
    desc = "desc"


class SearchScope(StrEnum):
    """Fields searchable via ?q=. The default (no `scope=` param) is
    every value below — passing one or more `scope=` values narrows
    the `q` match to just those.

    `title_` is suffixed to dodge the `str.title()` shadow; the wire
    value stays "title".

    `fulltext` matches the body text extracted from PDF attachments
    (see `attachments.tsv` GIN index + the worker pipeline). It's an
    EXISTS join rather than a column on items, so it short-circuits
    out of the OR'd metadata clauses.
    """

    title_ = "title"
    creators = "creators"
    abstract = "abstract"
    extra = "extra"
    fulltext = "fulltext"


async def _resolve_space(
    db: AsyncSession,
    user: User,
    slug: str,
    minimum: str = SPACE_ROLE_VIEWER,
) -> Space:
    result = await db.execute(select(Space).where(Space.slug == slug))
    space = result.scalar_one_or_none()
    await require_space_role(db, space, user.id, minimum, label="Space not found")
    assert space is not None  # require_space_role raises when it isn't
    return space


async def _resolve_item(
    db: AsyncSession,
    user: User,
    item_id: uuid.UUID,
    *,
    minimum: str = SPACE_ROLE_VIEWER,
    include_trashed: bool = False,
) -> Item:
    """Look up an item the caller is allowed to see. By default trashed
    items 404 — callers that operate on the trash (restore, permanent
    delete) opt in via include_trashed=True.

    `minimum` is the role the caller needs in the item's space: viewer to
    read, editor to change anything.
    """
    item = await db.get(Item, item_id)
    if item is None or (item.deleted_at is not None and not include_trashed):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Item not found")
    space = await db.get(Space, item.space_id)
    await require_space_role(db, space, user.id, minimum, label="Item not found")
    return item


@router.get("/api/spaces/{slug}/items", response_model=ListItemsResponse)
async def list_items(
    slug: str,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
    q: Annotated[str | None, Query(max_length=200)] = None,
    tag: Annotated[list[str] | None, Query()] = None,
    status_: Annotated[ItemStatus, Query(alias="status")] = ItemStatus.active,
    sort: Annotated[ItemSort, Query()] = ItemSort.updated,
    direction: Annotated[SortDirection, Query()] = SortDirection.desc,
    collection: Annotated[str | None, Query(max_length=64)] = None,
    scope: Annotated[list[SearchScope] | None, Query()] = None,
) -> dict[str, Any]:
    space = await _resolve_space(db, user, slug)
    stmt = select(Item).where(Item.space_id == space.id)
    if status_ is ItemStatus.active:
        stmt = stmt.where(Item.deleted_at.is_(None))
    elif status_ is ItemStatus.trashed:
        stmt = stmt.where(Item.deleted_at.is_not(None))
    # `all` skips the deleted_at predicate entirely.

    # Scope-priority ranking. When q is set, items with a title hit
    # come before items whose only match is in creators / abstract /
    # extra, and fulltext-only matches go last. Otherwise a query
    # like "002" (which matches hundreds of PDF bodies) buries the
    # handful of title hits past the first page and the user never
    # sees them. Built outside the q block so it's None when no
    # search is active.
    scope_rank = None
    if q and q.strip():
        # Case-insensitive substring match across selected metadata
        # fields. Creators are a JSONB array of objects, so we cast
        # to text and ilike the result — works for first/last/`name`
        # without a normalised people table. The `fulltext` scope
        # bolts on an EXISTS join against attachments.tsv (filled by
        # the extraction worker) so a PDF body match surfaces the
        # parent item alongside metadata hits.
        enabled = set(scope) if scope else set(SearchScope)
        cleaned = q.strip()
        needle = f"%{cleaned}%"
        clauses = []
        title_clause = Item.data["title"].astext.ilike(needle)
        creators_clause = cast(Item.data["creators"], String).ilike(needle)
        abstract_clause = Item.data["abstractNote"].astext.ilike(needle)
        extra_clause = Item.data["extra"].astext.ilike(needle)
        if SearchScope.title_ in enabled:
            clauses.append(title_clause)
        if SearchScope.creators in enabled:
            clauses.append(creators_clause)
        if SearchScope.abstract in enabled:
            clauses.append(abstract_clause)
        if SearchScope.extra in enabled:
            clauses.append(extra_clause)
        if SearchScope.fulltext in enabled:
            # Case-insensitive substring match, exactly like the
            # reader's in-PDF Ctrl-F find tool — a search for "Test"
            # picks up "Testing", "Latest", etc. The lexeme-based
            # tsvector approach stems away these substrings, which
            # surprised users (an item shows up under Ctrl-F but not
            # in the library search). The same matcher powers
            # /fulltext-hits so the listing and the expansion always
            # agree.
            text_needle = f"%{_escape_ilike(cleaned)}%"
            clauses.append(
                select(AttachmentPage.attachment_id)
                .join(
                    Attachment,
                    Attachment.id == AttachmentPage.attachment_id,
                )
                .where(
                    Attachment.item_id == Item.id,
                    AttachmentPage.text.ilike(text_needle, escape="\\"),
                )
                .exists()
            )
        if clauses:
            stmt = stmt.where(or_(*clauses))
            # Lower number = higher priority. Mirrors the order the
            # SPA already uses to bucket results into "Title hits" /
            # "Creator hits" / etc. sections.
            scope_rank = case(
                (title_clause, 0),
                (creators_clause, 1),
                (abstract_clause, 2),
                (extra_clause, 3),
                else_=4,
            )
        else:
            # All scopes disabled — by definition no matches.
            return {"items": [], "total": 0}
    if tag:
        # AND across multiple ?tag= params: each tag must be present.
        # Tags are matched by name (case-insensitive via the CITEXT
        # column) against the normalised `tags` table joined through
        # `item_tags`. Subquery-per-tag keeps the SQL straightforward
        # and the cardinality is small for any single user's library.
        for t in tag:
            t_clean = t.strip()
            if t_clean:
                stmt = stmt.where(
                    select(ItemTag.item_id)
                    .join(Tag, Tag.id == ItemTag.tag_id)
                    .where(
                        ItemTag.item_id == Item.id,
                        Tag.space_id == space.id,
                        Tag.name == t_clean,
                    )
                    .exists()
                )

    if collection:
        # Special-cased "unfiled" → items with no collection memberships.
        if collection == "unfiled":
            stmt = stmt.where(
                ~select(ItemCollection.item_id)
                .where(ItemCollection.item_id == Item.id)
                .exists()
            )
        else:
            try:
                cid = uuid.UUID(collection)
            except ValueError as e:
                raise HTTPException(
                    status.HTTP_400_BAD_REQUEST,
                    "collection must be a UUID or 'unfiled'",
                ) from e
            stmt = stmt.join(
                ItemCollection, ItemCollection.item_id == Item.id
            ).where(ItemCollection.collection_id == cid)

    # Sort. Title sort dips into the JSONB blob via data->>'title' and
    # is therefore not index-backed; build an expression index when the
    # row count makes it noticeable.
    sort_columns = {
        ItemSort.updated: Item.updated_at,
        ItemSort.created: Item.created_at,
        ItemSort.title_: Item.data["title"].astext,
        ItemSort.type_: Item.item_type,
    }
    primary = sort_columns[sort]
    primary = primary.asc() if direction is SortDirection.asc else primary.desc()
    # Stable order tie-breaker on id keeps pagination consistent across
    # rows that share the primary sort key. ``scope_rank`` (only set
    # during a search) leads so title hits surface above body-text
    # hits regardless of update time.
    if scope_rank is not None:
        stmt = stmt.order_by(scope_rank, primary, Item.id)
    else:
        stmt = stmt.order_by(primary, Item.id)

    # Total across all matches for this filter set — drives the
    # "X items" header in the SPA without it having to fetch every
    # page. ``order_by(None)`` strips the ORDER BY since count() over
    # a sub-select doesn't need ordering and Postgres rejects an
    # ORDER BY in some sub-select contexts otherwise.
    count_stmt = select(func.count()).select_from(
        stmt.order_by(None).subquery()
    )
    total = (await db.execute(count_stmt)).scalar_one()

    result = await db.execute(stmt.limit(limit).offset(offset))
    items = await _attach_collection_ids(db, list(result.scalars().all()))
    return {"items": items, "total": total}


@router.post(
    "/api/spaces/{slug}/items",
    response_model=ItemResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_item(
    slug: str,
    payload: ItemCreate,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
) -> dict[str, Any]:
    space = await _resolve_space(db, user, slug, SPACE_ROLE_EDITOR)
    item = Item(
        space_id=space.id,
        item_type=payload.item_type,
        data=payload.data,
        created_by=user.id,
    )
    db.add(item)
    await db.commit()
    await db.refresh(item)
    return (await _attach_collection_ids(db, [item]))[0]


@router.get("/api/items/{item_id}", response_model=ItemResponse)
async def get_item(
    item_id: uuid.UUID,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
) -> dict[str, Any]:
    item = await _resolve_item(db, user, item_id)
    return (await _attach_collection_ids(db, [item]))[0]


class FulltextHit(BaseModel):
    page_number: int
    snippet_html: str


class FulltextHitsAttachment(BaseModel):
    attachment_id: uuid.UUID
    filename: str
    hits: list[FulltextHit]


@router.get(
    "/api/items/{item_id}/fulltext-hits",
    response_model=list[FulltextHitsAttachment],
)
async def fulltext_hits(
    item_id: uuid.UUID,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
    q: Annotated[str, Query(min_length=1, max_length=200)],
) -> list[dict[str, Any]]:
    """Per-page snippet hits for the search expansion in LibraryPage.

    Substring/ILIKE-based — same matcher as the reader's in-PDF find
    tool, so the listing-level "this item matches" and the expansion
    "here are the pages" always agree. Snippet HTML is built in
    Python with html.escape + a single `<mark>` wrap so the client
    can render via dangerouslySetInnerHTML safely.
    """
    item = await _resolve_item(db, user, item_id)
    cleaned = q.strip()
    if not cleaned:
        return []

    att_rows = (
        await db.execute(
            select(Attachment.id, Attachment.filename).where(
                Attachment.item_id == item.id
            )
        )
    ).all()
    if not att_rows:
        return []

    text_needle = f"%{_escape_ilike(cleaned)}%"

    out: list[dict[str, Any]] = []
    for att_id, filename in att_rows:
        page_rows = (
            await db.execute(
                select(
                    AttachmentPage.page_number,
                    AttachmentPage.text,
                )
                .where(
                    AttachmentPage.attachment_id == att_id,
                    AttachmentPage.text.ilike(text_needle, escape="\\"),
                )
                .order_by(AttachmentPage.page_number.asc())
            )
        ).all()
        if not page_rows:
            continue
        hits: list[dict[str, Any]] = []
        for page_number, text in page_rows:
            snippet = _build_snippet(text, cleaned)
            if snippet is None:
                continue
            hits.append({"page_number": page_number, "snippet_html": snippet})
        if hits:
            out.append(
                {
                    "attachment_id": att_id,
                    "filename": filename,
                    "hits": hits,
                }
            )
    return out


@router.patch("/api/items/{item_id}", response_model=ItemResponse)
async def update_item(
    item_id: uuid.UUID,
    payload: ItemUpdate,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
) -> dict[str, Any]:
    item = await _resolve_item(db, user, item_id, minimum=SPACE_ROLE_EDITOR)
    if payload.item_type is not None:
        item.item_type = payload.item_type
    if payload.data is not None:
        item.data = payload.data
    await db.commit()
    await db.refresh(item)
    return (await _attach_collection_ids(db, [item]))[0]


@router.delete("/api/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_item(
    item_id: uuid.UUID,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
    permanent: Annotated[bool, Query()] = False,
) -> None:
    """Soft-delete an active item, or hard-delete a trashed one.

    Default behaviour mirrors a "move to trash" gesture — the row stays
    in the table with deleted_at set so the user can restore it. Pass
    ?permanent=true on a row that is already trashed to drop it for
    good (used by Empty Trash). Trying to permanent-delete an active
    item is rejected so a single accidental click can't take an item
    straight off the shelf.
    """
    item = await _resolve_item(
        db, user, item_id, minimum=SPACE_ROLE_EDITOR, include_trashed=permanent
    )
    if permanent:
        if item.deleted_at is None:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST,
                "Item must be trashed before it can be permanently deleted",
            )
        await db.delete(item)
    else:
        item.deleted_at = datetime.now(UTC)
    await db.commit()


@router.post("/api/items/{item_id}/restore", response_model=ItemResponse)
async def restore_item(
    item_id: uuid.UUID,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
) -> Item:
    item = await _resolve_item(
        db, user, item_id, minimum=SPACE_ROLE_EDITOR, include_trashed=True
    )
    if item.deleted_at is None:
        # No-op restore is fine to expose; just don't write.
        return item
    item.deleted_at = None
    await db.commit()
    await db.refresh(item)
    return item
